clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java

```diff
@@ -274,79 +269,18 @@ private void closeInternal(final Duration timeout) {
     }
 
     void cleanup() {
+        log.trace("Closing the consumer network thread");
+        Timer timer = time.timer(closeTimeout);
         try {
-            log.trace("Closing the consumer network thread");
-            Timer timer = time.timer(closeTimeout);
-            maybeAutocommitOnClose(timer);
             runAtClose(requestManagers.entries(), networkClientDelegate, timer);
-            maybeLeaveGroup(timer);
         } catch (Exception e) {
             log.error("Unexpected error during shutdown.  Proceed with closing.", e);
         } finally {
+            networkClientDelegate.awaitPendingRequests(timer);
```

Comment:
Network requests are tied to the `CompletableApplicationEvent`s, right? Can we just rely on the events to wait for their network I/O to complete via the `addAndGet()` method.?

Reply:
it seems like there's network poll in both runAtClose and here.  We really only need to poll it in a single place.  I wonder if we could just remove the polls in runAtClose() and try to send out the requests here.